### ISSUE TITLE
Crash on Windows 10 - Race condition with CrashHandler recreation and Enumerating GPU threads

### ISSUE THREAD

**Issue by lfantoniosi** [baldurk/renderdoc#2376]
## Description

RenderDoc official build v1.15 crashes on Windows 10 Enterprise after a while, the windows sessions needs to stay open for a day or two. It requires a fresh reboot to work again. Then I found out if I rebuild from source code the crash never happens and I discovered that the CrashHandler itself is the culprit.

Resuming short: you need to join all GPU enumerating threads BEFORE deleting the existing CrashHandler. Looks like after a while the enumerating GPU threads become slower and slower and a race condition causes the crash because the crash handler have been deleted and not yet recreated when it hits in another thread:

WrappedID3D11Device *wrappedDev = new WrappedID3D11Device(device, initParams); 

that itself will call for a CrashHandler that dies in a pure virtual call.

We are hitting it with remote desktop I don't know if is part of que equation, I cannot go to studio now until the pandemic ends
 
I think you need a join here:

```
    if(existing)
    {
      CrashHandler *crash = ((CrashHandler *)existing);
      m_PipeName = crash->m_PipeName;
      mem = crash->m_ExHandler->QueryRegisteredAppMemory();
      RDCLOG("Re-using crash-handling server %s", m_PipeName.c_str());
      SAFE_DELETE(existing);
    }
```

## Steps to reproduce

Leave the windows session open for a few days, running a remote session with Remote Desktop and then run RenderDoc official build (with CrashHandler active)

## Environment

It doesn't matter the version if I build myself with #define RENDERDOC_OFFICIAL_BUILD 1 I get the same issue

* RenderDoc version: 1.15 (but also found in 1.12)
* Operating System: Windows 10 Enterprise
* Graphics API: D3D11, D3D12, Vulkan, RTX2080 SUPER

It does not matter the driver or video card. A colleague got same issue on a GTX1080

**Comment by baldurk** [comment]
Thanks for digging into this. I've had one or two crash reports that in retrospect are presumably this bug, but I never got any repro steps and I couldn't work backwards from the minidump to see what went wrong. I'm surprised I haven't run into this myself, but I guess it depends entirely on how that race condition ends up working out.

**Comment by baldurk** [comment]
Looking at it in more detail I suspect it was maybe the crash handler recreate that was taking longer over time, rather than GPU enumeration. Either way, I was able to make it happen reliably with some judicious sleeps, and after rearranging the order of things so the recreate happens immediately before kicking off the GPU enumeration instead of immediately after it fixed the crash. I've also added some locking around recreating the handler just in case this happens somewhere else in future.
